handlers/mist_triggers.go

```diff
@@ -0,0 +1,196 @@
+		}
+		// Multiple pushes from RENDITION_PREFIX are in progress.
+		return
+	}
+	tracks := make(LiveTrackListTriggerJson)
+	if err := json.Unmarshal([]byte(encodedTracks), &tracks); err != nil {
+		Err(streamName, "LiveTrackListTriggerJson json decode error", err, w)
+		return
+	}
+	// Start push per each video track
+	info, err := d.StreamCache.Transcoding.Get(streamName)
+	if err != nil {
+		Err(streamName, "LIVE_TRACK_LIST unknown push source", err, w)
+		return
+	}
+	for i := range tracks { // i is generated name, not important, all info is contained in element
+		if tracks[i].Type != "video" {
+			// Only produce an rendition per each video track, selecting best audio track
+			continue
+		}
+		destination := fmt.Sprintf("%s/%s__%dx%d.ts?video=%d&audio=maxbps", info.UploadDir, streamName, tracks[i].Width, tracks[i].Height, tracks[i].Index) //.Id)
+		if err := d.MistClient.PushStart(streamName, destination); err != nil {
+			log.Printf("> ERROR push to %s %v", destination, err)
+		} else {
+			d.StreamCache.Transcoding.AddDestination(streamName, destination)
+		}
+	}
+}
+
+// TriggerPushEnd responds to PUSH_END trigger
+// This trigger is run whenever an outgoing push stops, for any reason.
+// This trigger is stream-specific and non-blocking. The payload for this trigger is multiple lines,
+// each separated by a single newline character (without an ending newline), containing data:
+//   push ID (integer)
+//   stream name (string)
+//   target URI, before variables/triggers affected it (string)
+//   target URI, afterwards, as actually used (string)
+//   last 10 log messages (JSON array string)
+//   most recent push status (JSON object string)
+func (d *MistCallbackHandlersCollection) TriggerPushEnd(w http.ResponseWriter, req *http.Request, payload []byte) {
+	lines := strings.Split(strings.TrimSuffix(string(payload), "\n"), "\n")
+	if len(lines) != 6 {
+		errors.WriteHTTPBadRequest(w, "Bad request payload", fmt.Errorf("unknown payload '%s'", string(payload)))
+		return
+	}
+	// stream name is the second line in the Mist Trigger payload
+	streamName := lines[1]
+	destination := lines[2]
+	actualDestination := lines[3]
+	pushStatus := lines[5]
+	if yes, _ := isTranscodeStream(streamName); yes {
+		d.TranscodingPushEnd(w, req, streamName, destination, actualDestination, pushStatus)
+	} else {
+		d.SegmentingPushEnd(w, req, streamName)
+	}
+}
+
+func (d *MistCallbackHandlersCollection) TranscodingPushEnd(w http.ResponseWriter, req *http.Request, streamName, destination, actualDestination, pushStatus string) {
+	info, err := d.StreamCache.Transcoding.Get(streamName)
+	if err != nil {
+		Err(streamName, "unknown push source", err, w)
+		return
+	}
+	isBeingProcessed := false
+	for _, existing := range info.Destinations {
+		if existing == destination {
+			isBeingProcessed = true
+			break
+		}
+	}
+	if !isBeingProcessed {
+		Err(streamName, "missing from our books", nil, w)
```

Comment:
Maybe `"stream not found in cache"`? Not sure we use `"books"` in our domain language.